src/service/resources-impl.js

```diff
@@ -468,80 +483,105 @@ export class ResourcesImpl {
       this.ampdoc.getVisibilityState() != VisibilityState.PRERENDER ||
       resource.prerenderAllowed();
 
-    if (buildingEnabled && shouldBuildResource) {
+    const isCloseEnoughToViewport = this.buildWhenCloseToViewport_
+      ? ignoreQuota ||
+        resource.isBuildRenderBlocking() ||
+        resource.renderOutsideViewport() ||
+        (this.isIdle_() && resource.idleRenderOutsideViewport())
+      : true;
+
+    if (
+      buildingEnabled &&
+      !isBuiltOrBuilding &&
```

Comment:
Again, careful here with changing the order the conditional. Before I fixed one of the race conditions in resource.js, calling build on already-built elements (!) was a happy escape hatch for that race condition.